From: Daniel Kahn Gillmor <dkg@fifthhorseman.net>
Date: Thu, 4 Oct 2018 21:43:38 -0500
Subject: Use distinct GnuPG homedirs for each test

As thunderbird has become more capable of running tests
asynchronously, having multiple tests which use the same GnuPG homedir
can cause problems with the test suite.

In particular, if test A is launched, and completes successfully while
test B is being launched, then withTestGpgHome from test A might
delete the homedir while test B has already begun.

If this happens, it can result in some weird race conditions.

This fix doesn't guarantee that the homedir is distinct (perhaps we
should change initializeGpgHome() to createUnique instead of just
create to make the directory?), but it makes a collision very
improbable.
---
 package/tests/gpgAgent-test.js | 2 +-
 package/tests/testHelper.js    | 8 ++++++--
 2 files changed, 7 insertions(+), 3 deletions(-)

diff --git a/package/tests/gpgAgent-test.js b/package/tests/gpgAgent-test.js
index e5a2acb..befec9e 100644
--- a/package/tests/gpgAgent-test.js
+++ b/package/tests/gpgAgent-test.js
@@ -212,7 +212,7 @@ test(withTestGpgHome(withEnigmail(function shouldGetGpgHomeDir() {
   let homedirExpected = osUtils.OS.Path.join(EnigmailFiles.getTempDir(), ".gnupgTest");
 
   let homeDir = EnigmailGpgAgent.getGpgHomeDir();
-  Assert.equal(homedirExpected, homeDir);
+  Assert.equal(homedirExpected, homeDir.substr(0, homedirExpected.length));
 })));
 
 // getHomedirFromParam
diff --git a/package/tests/testHelper.js b/package/tests/testHelper.js
index d4e0070..e533acf 100644
--- a/package/tests/testHelper.js
+++ b/package/tests/testHelper.js
@@ -11,6 +11,8 @@
 const osUtils = {};
 Components.utils.import("resource://gre/modules/osfile.jsm", osUtils);
 Components.utils.import("resource://gre/modules/FileUtils.jsm", osUtils);
+Components.utils.import("resource://enigmail/rng.jsm"); /*global EnigmailRNG: false */
+Components.utils.import("resource://enigmail/core.jsm"); /*global EnigmailCore: false */
 
 var TestHelper = {
   loadDirectly: function(name) {
@@ -54,7 +56,7 @@ var TestHelper = {
 
   initalizeGpgHome: function() {
     component("enigmail/files.jsm");
-    var homedir = osUtils.OS.Path.join(EnigmailFiles.getTempDir(), ".gnupgTest");
+    var homedir = osUtils.OS.Path.join(EnigmailFiles.getTempDir(), ".gnupgTest" + EnigmailRNG.generateRandomString(8));
     var workingDirectory = new osUtils.FileUtils.File(homedir);
     if (!workingDirectory.exists()) {
       workingDirectory.create(Components.interfaces.nsIFile.DIRECTORY_TYPE, 448);
@@ -88,7 +90,9 @@ var TestHelper = {
     var environment = Components.classes["@mozilla.org/process/environment;1"].getService(Components.interfaces.nsIEnvironment);
 
     environment.set("GNUPGHOME", workingDirectory.path);
-    return homedir;
+    if (EnigmailCore.getEnvList() !== null)
+      EnigmailCore.setEnvVariable("GNUPGHOME", workingDirectory.path);
+   return homedir;
   },
 
   removeGpgHome: function(homedir) {
